Cargo.toml

```diff
@@ -29,5 +29,10 @@ tracing-subscriber = { version = "0.3.17", features = ["env-filter"] }
 url = "2.3.1"
 urlencoding = "2.1.2"
 
+[profile.dev]
+lto = "thin"
```

Comment:
Thanks!

What are the pros and cons of enabling LTO in debug builds during development?

Reply:
The only con will be a slight increase in compile times, but for a project this size it's marginal.

Reply:
Hi Jonathan, thanks for your interest in the project!

From my perspective, there's no need to enable LTO in the dev profile; we want faster builds at the expense of lower performance. The current performance of debug builds is fine for development in my experience.

Generally, my preference is to stick with the rust/cargo defaults unless there is a clear measured performance benefit. The last time I looked into this it seemed that `gzip` was the main bottleneck when processing. I don't have the bandwidth to run more performance tests right now, but my guess is that LTO would not provide much benefit with our current workload.

Reply:
It is worth trying at least, it should not harm in Release. Debug builds don't need it, the development speed (read: developer's time) is very important.

Reply:
I don't have the necessary pieces to test the full pipeline, but this would seem to consistently shave off a second of user CPU on a basic PBF file:

```
# Non-LTO

    Finished release [optimized] target(s) in 1m 06s

$ hyperfine -m 1000 './target/release/om-wikiparser get-tags greater-london-latest.osm.pbf'
Benchmark 1: ./target/release/om-wikiparser get-tags greater-london-latest.osm.pbf
  Time (mean ± σ):     457.4 ms ±  17.9 ms    [User: 5686.1 ms, System: 1580.0 ms]
  Range (min … max):   416.3 ms … 536.4 ms    1000 runs


# LTO

    Finished release [optimized] target(s) in 1m 39s

$ hyperfine -m 1000 './target/release/om-wikiparser get-tags greater-london-latest.osm.pbf'
Benchmark 1: ./target/release/om-wikiparser get-tags greater-london-latest.osm.pbf
  Time (mean ± σ):     468.6 ms ±  22.1 ms    [User: 4696.6 ms, System: 1674.1 ms]
  Range (min … max):   412.3 ms … 552.7 ms    1000 runs
```

Reply:
I'm a bit confused by that output. The wall-clock time difference between the two is ~10ms, half the standard deviation with or without LTO.
The `get-tags` subcommand is multi-threaded, and the User/System times [are the sum of all threads](https://www.github.com/sharkdp/hyperfine/issues/515#issuecomment-1135695373). Unfortunately hyperfine [doesn't report the deviation, range, or per-run times of the User/System figures](https://www.github.com/sharkdp/hyperfine/issues/660). Forcing a single worker thread with `OM_POOL_THREADS` or `-t` might give a better idea of the impact.

Either way, the vast majority of time in real usage is spent in the `get-articles` subcommand. I think a decent way to profile that would be:
1. Find the general ratio of matched/unmatched articles in the wikipedia dumps.
2. Make a small subset of a dump with that representative ratio and store it decompressed.
3. Profile for "thin"/"fat"/false with the subset, and maybe also with only no matches/all matches.

------

My understanding of "thin" vs "fat"/true:
- "fat" requires a fully serialized step after linking, optimizes across compilation units. Large time and memory requirements.
- "thin" is a newer method that is more parallelizable with less overhead, but generally almost as effective as "fat".

Does that match your understanding?

> It is worth trying at least, it should not harm in Release

I remember reading several anecdotes that LTO can hurt performance in some cases, especially with "small" programs, but I'm having trouble finding those again. There have been times where [it interacts poorly with other options](https://www.github.com/rust-lang/rust/issues/47745).
I think it's best to treat it like other optimization settings and verify that it helps, at the very least to make sure we're using the correct settings.